extensions/parquet/table/src/main/java/io/deephaven/parquet/table/pagestore/OffsetIndexBasedColumnChunkPageStore.java

```diff
@@ -0,0 +1,147 @@
+            }
+        }
+        fixedPageSize = isPageSizeFixed ? firstPageSize : PAGE_SIZE_NOT_FIXED;
+    }
+
+    /**
+     * Binary search in offset index to find the page number that contains the row. Logic duplicated from
+     * {@link Arrays#binarySearch(long[], long)} to use the offset index.
+     */
+    private static int findPageNumUsingOffsetIndex(final OffsetIndex offsetIndex, final long row) {
+        int low = 0;
+        int high = offsetIndex.getPageCount() - 1;
+
+        while (low <= high) {
+            final int mid = (low + high) >>> 1;
+            final long midVal = offsetIndex.getFirstRowIndex(mid);
+
+            if (midVal < row)
+                low = mid + 1;
+            else if (midVal > row)
+                high = mid - 1;
+            else
+                return mid; // key found
+        }
+        return -(low + 1); // key not found.
+    }
+
+    private ChunkPage<ATTR> getPage(final int pageNum) {
+        if (pageNum < 0 || pageNum >= numPages) {
+            throw new IllegalArgumentException("pageNum " + pageNum + " is out of range [0, " + numPages + ")");
+        }
+        PageCache.IntrusivePage<ATTR> page;
+        PageState pageState = pageStates.get(pageNum);
+        if (pageState == null) {
+            pageState = pageStates.updateAndGet(pageNum, p -> p == null ? new PageState() : p);
+        }
+        if (pageState.pageRef == null || (page = pageState.pageRef.get()) == null) {
+            synchronized (pageState) {
+                // Make sure no one materialized this page as we waited for the lock
+                if (pageState.pageRef == null || (page = pageState.pageRef.get()) == null) {
+                    final ColumnPageReader reader = columnPageDirectAccessor.getPageReader(pageNum);
+                    try {
+                        page = new PageCache.IntrusivePage<>(toPage(offsetIndex.getFirstRowIndex(pageNum), reader));
+                    } catch (final IOException except) {
+                        throw new UncheckedIOException(except);
+                    }
+                    pageState.pageRef = new WeakReference<>(page);
+                }
+            }
+        }
+        pageCache.touch(page);
+        return page.getPage();
+    }
+
+    @NotNull
+    @Override
+    public ChunkPage<ATTR> getPageContaining(@NotNull final FillContext fillContext, long row) {
+        row &= mask();
+        Require.inRange(row, "row", numRows(), "numRows");
+
+        int pageNum;
+        if (fixedPageSize == PAGE_SIZE_NOT_FIXED) {
+            pageNum = findPageNumUsingOffsetIndex(offsetIndex, row);
+            if (pageNum < 0) {
+                pageNum = -2 - pageNum;
+            }
+        } else {
+            pageNum = (int) (row / fixedPageSize);
+            if (pageNum >= numPages) {
+                // This can happen if the last page is of different size from rest of the pages.
+                // We have already checked that row is less than numRows.
```

Comment:
I'm trying to think if this is actually true and could happen.

Let's say we have `fixedPageSize = 100` and `numPages = 3`, and the last page only has 1 row; so, `numRows = 100 * 2 + 1 = 201`.

```
(int)(x / 100)
0, x=[0, 100)
1, x=[100, 200)
2, x=[200, 201)
```

At no point did we ever get back `>=3`.